src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveRenderedService.cs

```diff
@@ -0,0 +1,127 @@
+        EnsureArg.IsNotNull(retrieveConfiguration?.Value, nameof(retrieveConfiguration));
+        EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        _instanceStore = instanceStore;
+        _blobDataStore = blobDataStore;
+        _dicomRequestContextAccessor = dicomRequestContextAccessor;
+        _retrieveConfiguration = retrieveConfiguration?.Value;
+        _recyclableMemoryStreamManager = recyclableMemoryStreamManager;
+        _logger = logger;
+    }
+
+    public async Task<RetrieveRenderedResponse> RetrieveRenderedImageAsync(RetrieveRenderedRequest request, CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(request, nameof(request));
+        var partitionKey = _dicomRequestContextAccessor.RequestContext.GetPartitionKey();
+
+        try
+        {
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+
+            AcceptHeader returnHeader = GetValidRenderAcceptHeader(request.AcceptHeaders);
+
+            // this call throws NotFound when zero instance found
+            InstanceMetadata instance = (await _instanceStore.GetInstancesWithProperties(
+                ResourceType.Instance, partitionKey, request.StudyInstanceUid, request.SeriesInstanceUid, request.SopInstanceUid, cancellationToken)).First();
+
+            FileProperties fileProperties = await RetrieveHelpers.CheckFileSize(_blobDataStore, _retrieveConfiguration.MaxDicomFileSize, instance, cancellationToken);
+            Stream stream = await _blobDataStore.GetFileAsync(instance.VersionedInstanceIdentifier, cancellationToken);
+            DicomFile dicomFile = await DicomFile.OpenAsync(stream, FileReadOption.ReadLargeOnDemand);
+            DicomPixelData dicomPixelData = dicomFile.GetPixelDataAndValidateFrames(new[] { request.FrameNumber });
+
+            DicomImage dicomImage = new DicomImage(dicomFile.Dataset);
+            using var img = dicomImage.RenderImage(request.FrameNumber);
+            using var sharpImage = img.AsSharpImage();
+            MemoryStream resultStream = _recyclableMemoryStreamManager.GetStream();
+            await sharpImage.SaveAsJpegAsync(resultStream, new SixLabors.ImageSharp.Formats.Jpeg.JpegEncoder(), cancellationToken: cancellationToken);
+            resultStream.Position = 0;
+
+            string outputContentType = returnHeader.MediaType.ToString();
+
+            sw.Stop();
+            _logger.LogInformation("Render from dicom to {OutputContentType}, uncompressed frame size was {UncompressedFrameSize}, output frame size is {OutputFrameSize} and took {ElapsedMilliseconds} ms", outputContentType, dicomPixelData.UncompressedFrameSize, resultStream.Length, sw.ElapsedMilliseconds);
+
+            return new RetrieveRenderedResponse(resultStream, resultStream.Length, outputContentType);
+        }
+        catch (IndexOutOfRangeException e)
```

Comment:
I understand RenderImage throws IndexOutofRangeException when frame doesn't exists, but it is not specific. Is there any specific error code or message that it gives?

Reply:
Can we check if Dataset contains Pixeldata tag? If not return immediately, we can avoid creating Dicomimage and calling RenderImage

Reply:
as discussed offline looks like fodicom does these optimizations for us so we will not need to do it